osutil/user/getent.go

```diff
@@ -0,0 +1,127 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package user
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"os/exec"
+	"strings"
+	"unicode"
+)
+
+func getEnt(params ...string) ([]byte, error) {
+	var outBuf bytes.Buffer
+	var errBuf bytes.Buffer
+
+	cmd := exec.Command("getent", params...)
+	cmd.Stdin = nil
+	cmd.Stdout = &outBuf
+	cmd.Stderr = &errBuf
+
+	if err := cmd.Run(); err != nil {
+		return nil, fmt.Errorf("getent returned an error: %q", errBuf.Bytes())
+	}
+
+	return outBuf.Bytes(), nil
+}
+
+// lookupFromGetent calls getent, parses and filters its output
+// The component at `index` will need to match `expectedValue`.
+// If `isKey`, then `expectedValue` will also be passed as parameter
+// to getent along `database`. `numComponents` should be 4 for groups
+// and 7 for users.
+func lookupFromGetent(database string, index int, expectedValue string, isKey bool, numComponents int) ([]string, error) {
+	params := []string{database}
+	if isKey {
+		params = append(params, expectedValue)
+	}
+	buf, err := getEnt(params...)
+	if err != nil {
+		return nil, err
+	}
+	scanner := bufio.NewScanner(bytes.NewReader(buf))
+	for scanner.Scan() {
+		components := strings.SplitN(scanner.Text(), ":", numComponents)
+		if len(components) != numComponents {
+			continue
+		}
+
+		if components[index] != expectedValue {
+			continue
+		}
+
+		return components, nil
+	}
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+	return nil, nil
+}
+
+func isNumeric(value string) bool {
+	for _, c := range value {
+		return unicode.IsDigit(c)
```

Comment:
I don't think we want to use this. This will return true for roman numerals, fractions and other weird stuff: https://cs.opensource.google/go/go/+/refs/tags/go1.23.1:src/unicode/digit.go

Separately this is a misnomer. Unicode has deifnitions for digits and number that are different. There's also `unicode.IsNumber` which we are not calling here.

Reply:
I don't think it's a big issue if we simply check that `c >= '0' && c <= '9'`?

Reply:
This is what I have changed to.